index/scorch/introducer.go

```diff
@@ -238,14 +257,32 @@ func (s *Scorch) introduceMerge(nextMerge *segmentMerge) {
 		}
 	}
 
-	// put new segment at end
-	newSnapshot.segment = append(newSnapshot.segment, &SegmentSnapshot{
-		id:         nextMerge.id,
-		segment:    nextMerge.new, // take ownership for nextMerge.new's ref-count
-		deleted:    newSegmentDeleted,
-		cachedDocs: &cachedDocs{cache: nil},
-	})
-	newSnapshot.offsets = append(newSnapshot.offsets, running)
+	// before the newMerge introduction, need to clean the newly
+	// merged segment wrt the current root segments, hence
+	// applying the obsolete segment contents to newly merged segment
+	for segID, ss := range nextMerge.old {
+		obsoleted := ss.DocNumbersLive()
+		if obsoleted != nil {
+			obsoletedIter := obsoleted.Iterator()
+			for obsoletedIter.HasNext() {
+				oldDocNum := obsoletedIter.Next()
+				newDocNum := nextMerge.oldNewDocNums[segID][oldDocNum]
+				newSegmentDeleted.Add(uint32(newDocNum))
+			}
+		}
+	}
+
+	if nextMerge.new != nil &&
+		nextMerge.new.Count() > newSegmentDeleted.GetCardinality() {
```

Comment:
i don't understand this condition, we're checking the size of the new segment, with the number of newly deleted items?  is this to catch the case that by the time the merge happened, all of the docs are actually now deleted (by doing these additional checks) and now there is no point in adding it?

if so, i'm wondering if there is a clearer way to express this, i had to sit and really think about the math to even imagine thats what it was doing

Reply:
Yes. The intent is to capture the case you mentioned. What if all the docs in the newly merged segment got deleted by the time we reach here. Then no point in adding. I agree that these may be confusing.. Shall we add some comments explaining the same at line:274?